### PR TITLE
PLAT-4524 Add pre-commit validations for terraform

### PR DESCRIPTION
**Pull request by miguelhar**
Adding pre-commit hooks:

  - id: check-merge-conflict
  - id: end-of-file-fixer
  - id: no-commit-to-branch
  - id: check-case-conflict
  - id: terraform_docs
  - id: terraform_fmt
  - id: terraform_tflint
  - id: terraform_validate
  - id: checkov
  - id: terraform_tfsec
  - id: terrascan

### CODE REVIEW

File: .terraform.lock.hcl [hunk 1]
```diff
@@ -0,0 +1,59 @@
+# This file is maintained automatically by "terraform init".
```
Comment (steved) [inline .terraform.lock.hcl:1]: do we want this committed?

File: main.tf [hunk 2]
```diff
@@ -289,12 +289,12 @@ resource "google_compute_firewall" "iap-tcp-forwarding" {
   }
 
   source_ranges = var.allowed_ssh_ranges
-  target_tags   = ["iap-tcp-forwarding-allowed"]
+  target_tags   = ["iap_tcp_forwarding-allowed"]
```
Comment (steved) [inline main.tf:292]: Was this caught by linting too? I feel like we should decide on a tag format (dashes or underscore) independent of the terraform resource standard

File: terraform.tf [hunk 3]
```diff
@@ -0,0 +1,17 @@
+terraform {
```
Comment (steved) [inline terraform.tf:1]: I believe it's @fraenkel's position that this is solely the responsibility of the consumer of this module

File: .terraform.lock.hcl (hunk 1 above)
Comment (miguelhar) [reply .terraform.lock.hcl:1]: Thats a great question, it boils down to 2 things.

1. How we want to react to new versions.
2. Increases the validity of the providers.

I am partial to committing it, because of #1 and the determinism in the version/sha explicit declaration.

more discussion [here](https://discuss.hashicorp.com/t/what-is-the-best-practise-with-terraform-lock-hcl-in-modules-commit-or-not/28648)

File: terraform.tf (hunk 3 above)
Comment (miguelhar) [reply terraform.tf:1]: 1. The issue is that the resources declaration is somewhat tied to a particular version of the provider,  there were fields in the resources declaration that are not backwards compatible with newer version, 
example.
the `location/zone` parameter  changed from one version of the provider to the next.

```
resource "google_filestore_instance" "instance" {
  zone = "us-central1-b"
...
```
after 3.9.0
```
resource "google_filestore_instance" "instance" {
  location = "us-central1-b"
...
}
```

2. It was caught by `tflint`
```
Warning: Missing version constraint for provider "google-beta" in "required_providers" (terraform_required_providers)

on [gcr.tf](http://gcr.tf/) line 1:

1: provider "google-beta" {

Reference: https://github.com/terraform-linters/tflint/blob/v0.35.0/docs/rules/terraform_required_providers.md

Warning: Missing version constraint for provider "google" in "required_providers" (terraform_required_providers)

on [main.tf](http://main.tf/) line 17:

17: provider "google" {

Reference: https://github.com/terraform-linters/tflint/blob/v0.35.0/docs/rules/terraform_required_providers.md

Warning: Missing version constraint for provider "random" in "required_providers" (terraform_required_providers)

on [main.tf](http://main.tf/) line 26:

26: resource "random_uuid" "id" {}

Reference: https://github.com/terraform-linters/tflint/blob/v0.35.0/docs/rules/terraform_required_providers.md
```

3. google official gke terraform module for [reference](https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/master/versions.tf)

In conclusion the absence of the version pinning tasks the customer with the burden of `figuring` out on which version the module "works".

File: main.tf (hunk 2 above)
Comment (miguelhar) [reply main.tf]: No, this was unintentional.

File: main.tf [hunk 4]
```diff
@@ -242,7 +242,7 @@ resource "google_container_node_pool" "node_pools" {
     }
 
     tags = [
-      "iap-tcp-forwarding-allowed",
+      "iap_tcp_forwarding-allowed",
```
Comment (Secretions) [inline main.tf:245]: I guess this was underscored to match the resource name, but since this is an admin-visible tag do we really want that to be the case? I feel like its formatting should be geared toward the end-user rather than the matching the terraform code.

Comment (Secretions) [reply main.tf:245]: I guess this is the same topic as one of steve's comments, but just a different tag?

**Comment by johnjoo1** [comment]
I have no substantive comment on this PR. Just wanted to tell @Secretions and @steved that I miss you guys! Hope you're doing well!

File: README.md [hunk 5]
```diff
@@ -64,3 +64,96 @@ It may be possible to lower the "admin" privilage levels to a "creator" level if
...
 
 Please submit any feature enhancements, bug fixes, or ideas via pull requests or issues.
+
+# Terraform Docs
+<!-- BEGINNING OF PRE-COMMIT-TERRAFORM DOCS HOOK -->
+## Requirements
+
+| Name | Version |
+|------|---------|
+| <a name="requirement_terraform"></a> [terraform](#requirement\_terraform) | >= 0.12 |
```
Comment (fraenkel) [inline README.md:74]: this is wrong, we are using 1.x.y

File: main.tf [hunk 6]
```diff
@@ -279,7 +279,7 @@ resource "google_container_node_pool" "node_pools" {
 }
 
 # https://cloud.google.com/iap/docs/using-tcp-forwarding
-resource "google_compute_firewall" "iap-tcp-forwarding" {
+resource "google_compute_firewall" "iap_tcp_forwarding" {
```
Comment (fraenkel) [inline main.tf:282]: if we claim any backward compatibility, we cannot do renames.

File: tests/.terraform.lock.hcl [hunk 7]
```diff
@@ -0,0 +1,60 @@
+# This file is maintained automatically by "terraform init".
```
Comment (fraenkel) [inline tests/.terraform.lock.hcl:1]: we should not be including this either.
This is a test, so it should track whatever is current given that reflects whatever anyone will use at any given moment.

File: variables.tf [hunk 8]
```diff
@@ -176,11 +180,23 @@ variable "node_pools" {
 }
 
 variable "node_pool_overrides" {
+  description = "Param override for var.node_pools"
+  type = map(
+    object({
```
Comment (fraenkel) [inline variables.tf:185]: I don't believe this is correct. This limits the overrides to just the 3 pools.
The node_pools can have any number of pools of which 3 have default values.

File: variables.tf [hunk 9]
```diff
@@ -176,11 +180,23 @@ variable "node_pools" {
...
+      gpu      = map(string)
+      platform = map(string)
+    })
+  )
   default = {}
 }
 
 variable "namespaces" {
-  type        = object({ platform = string, compute = string })
+  type = object(
```
Comment (fraenkel) [inline variables.tf:195]: what caused the formatting change? Are you using tf 1.x?

File: terraform.tf (hunk 3 above)
Comment (miguelhar) [reply terraform.tf:1]: I would agree if all the modules are declared and maintained on the same repo but not on remotely referenced  modules, which is the goal and the standard(can be seen on aws/google modules `versions.tf`)

from the footprint of the implicit providers block
```
Note: Only provider configurations are inherited by child modules, not provider source or version requirements. Each module must [declare its own provider requirements](https://www.terraform.io/language/providers/requirements). This is especially important for non-HashiCorp providers.
```

File: README.md (hunk 5 above)
Comment (miguelhar) [reply README.md:74]: I took this from the `test` requirements. will update

File: main.tf (hunk 6 above)
Comment (miguelhar) [reply main.tf:282]: I raised the same point, but was directed that the module was not used by customers. This will require a major version bump

File: tests/.terraform.lock.hcl (hunk 7 above)
Comment (miguelhar) [reply tests/.terraform.lock.hcl:1]: I dont feel strongly on checking this file but, we are not tracking what is current. all this file does is provide the available `sha` for the specified versions.

File: variables.tf (hunk 8 above)
Comment (miguelhar) [reply variables.tf:185]: Should we update this to account for the x number of pools?
```
initial_node_count = local.node_pools.platform.initial_count + local.node_pools.compute.initial_count + local.node_pools.gpu.initial_count
```

File: tests/.terraform.lock.hcl (hunk 7 above)
Comment (fraenkel) [reply tests/.terraform.lock.hcl:1]: we don't need to... we tend to run a deploy against this often enough that any break would be caught.
If we really care, we can add a weekly CI check to guarantee it always works but that would mean we actually get notified on failure and do something about it.
Tracking means little if we aren't pinning the versions.

File: variables.tf (hunk 8 above)
Comment (fraenkel) [reply variables.tf:185]: We shouldn't have to. It is there to handle https://github.com/hashicorp/terraform-provider-google/issues/6901. We have set remove_default_pool = true which causes them to be destroyed anyhow. It is really there to size the control plane which is now multi-AZ. Prior it was single AZ and a scale up caused an outage.

File: variables.tf (hunk 9 above)
Comment (miguelhar) [reply variables.tf]: Updated the format, `tf_fmt` is happy with it.

File: terraform.tf (hunk 3 above)
Comment (miguelhar) [reply terraform.tf]: @fraenkel note the goal is to comply with industry standards and publish this module on the terraform registry.

File: tests/.terraform.lock.hcl (hunk 7 above)
Comment (fraenkel) [reply tests/.terraform.lock.hcl]: the checksum always occurs since its using the registry to download and verify.
the lock makes us use the version we did the initial `init` with. Unless we do an `init -upgrade` we will never be testing against them.

File: terraform.tf (hunk 3 above)
Comment (fraenkel) [reply terraform.tf]: So even tflint has a comment here:
https://github.com/terraform-linters/tflint/blob/master/docs/rules/terraform_required_providers.md#how-to-fix
```
This approach is now discouraged, particularly for child modules.
```

Comment (miguelhar) [reply terraform.tf]: child modules are not used here.

Comment (fraenkel) [reply terraform.tf]: it is given its a child module to platform-apps.

Comment (miguelhar) [reply terraform.tf]: What I meant was that this module does not have any child modules. Also this module is not considered a child module to platform apps unless platform apps is another module.

[child-modules](https://www.terraform.io/language/modules#child-modules)
```
A module that has been called by another module is often referred to as a child module.
```

Comment (fraenkel) [reply terraform.tf]: Everything is a module, the top level is called the root module.
https://www.terraform.io/language/modules#the-root-module

The point is that child modules are invoked via a module block which is how this is invoked.

Comment (fraenkel) [reply terraform.tf]: The issue with provider versions is that if we aren't careful we are going to have 2 sets, one from platform-apps and one from the module which creates a total mess. Again our use case is that everything is driven from platform-apps which means it should be defining the provider versions and the providers used by modules.

Comment (miguelhar) [reply terraform.tf]: if we want a module that only works with platform-apps then we we should stick it in a folder on its repo. But this is public published terraform module and as is any client that tries to consume it needs to figure out the right versions in order to use it(not only the providers but the terraform version), again see the use-case I noted.

Comment (miguelhar) [reply terraform.tf]: we can have different version constraints as long as it produces a union. As for 4.x, It was not part of the scope of the story, but we can add it.

Comment (fraenkel) [reply terraform.tf]: see how easy it is to add otherwise its just more work back and forth.

**Review by fraenkel — APPROVED** [review]
determine if we can more to 4.x with minimal changes
otherwise its good to go

File: main.tf [hunk 10]
```diff
@@ -110,6 +110,8 @@ resource "google_filestore_instance" "nfs" {
 }
 
 resource "google_container_cluster" "domino_cluster" {
+  provider = google-beta
```
Comment (fraenkel) [inline main.tf:113]: why the switch to google-beta?

**Comment by miguelhar** [comment]
Creating branch from repo since fork integration is not done
